File: csharp/main/TypeKind.cs
using System;

namespace Deltix.Luminary
{
	public enum TypeKind
	{
		Type,
		Boolean,
		Int8,
		UInt8,
		Int16,
		UInt16,
		Int32,
		UInt32,
		Int64,
		UInt64,
		Float32,
		Float64,
		Decimal,
		Text,
		Data,
		Timestamp,
		Duration,
		Time,
		Date,
		// ReSharper disable once InconsistentNaming
		UUID,
		Nullable,
		List,
		Map,
		Action,
		Function,
		Enumeration,
		Interface,
		Class,
		Decorator
	}

	public static class TypeKindExtensions
	{
		public static IntegralType ToIntegralType(this TypeKind type)
		{
			switch (type)
			{
				case TypeKind.Int8:
				case TypeKind.UInt8:
				case TypeKind.Int16:
				case TypeKind.UInt16:
				case TypeKind.Int32:
				case TypeKind.UInt32:
				case TypeKind.Int64:
				case TypeKind.UInt64:
					return (IntegralType)(Int32)type;

				default:
					throw new InvalidOperationException($"Type {type} is not integral.");
			}
		}
	}
}

File: java/src/main/java/deltix/luminary/TypeKind.java
package deltix.luminary;

public enum TypeKind {
    TYPE(0),
    BOOLEAN(1),
    INT8(2),
    UINT8(3),
    INT16(4),
    UINT16(5),
    INT32(6),
    UINT32(7),
    INT64(8),
    UINT64(9),
    FLOAT32(10),
    FLOAT64(11),
    DECIMAL(12),
    TEXT(13),
    DATA(14),
    TIMESTAMP(15),
    DATE(16),
    TIME(17),
    DURATION(18),
    UUID(19),

    NULLABLE(100),
    LIST(101),
    ACTION(102),
    FUNCTION(103),
    MAP(104),
    SET(105),

    ENUMERATION(1000),
    INTERFACE(1001),
    CLASS(1002),
    DECORATOR(1003);

    private final int number;

    TypeKind(int number) {
        this.number = number;
    }

    public int getNumber() {
        return number;
    }

    public IntegralType toIntegralType() {
        switch (this) {
            case INT8:
                return IntegralType.INT8;

            case UINT8:
                return IntegralType.UINT8;

            case INT16:
                return IntegralType.INT16;

            case UINT16:
                return IntegralType.UINT16;

            case INT32:
                return IntegralType.INT32;

            case UINT32:
                return IntegralType.UINT32;

            case INT64:
                return IntegralType.INT64;

            case UINT64:
                return IntegralType.UINT64;

            default:
                throw new IllegalStateException("Type '" + this + "' is not integral.");
        }
    }
}
